### PR TITLE
EVM decode

### PR DESCRIPTION
**Pull request by nbgl**

### CODE REVIEW

File: evm/src/cpu/columns.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+pub const IS_INVALID_18: usize = IS_INVALID_17 + 1;
+pub const IS_INVALID_19: usize = IS_INVALID_18 + 1;
+pub const IS_INVALID_20: usize = IS_INVALID_19 + 1;
+// An instruction is invalid if _any_ of the above flags is 1.
+
+pub const START_INSTRUCTION_FLAGS: usize = IS_STOP;
+pub const END_INSTRUCTION_FLAGS: usize = IS_INVALID_20 + 1;
+
+// If CPU cycle: the opcode, broken up into bits.
+// **big-endian** order
```
Comment (dlubarov) [inline evm/src/cpu/columns.rs:123]: FWIW Plonky2 uses little-endian elsewhere, for bignums and bitwise stuff (like `split_le`). Not a big deal if you think it's nicer / more standard in this context though.

File: evm/src/cpu/decode.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    // Now check that they match the opcode.
+    let opcode = lv[columns::OPCODE];
+    yield_constr.constraint(cycle_filter * (opcode - top_bits[8]));
+
+    // Check that the instruction flags are valid.
+    // First, check that they are all either 0 or 1.
+    for &flag in lv
+        .iter()
+        .take(columns::END_INSTRUCTION_FLAGS)
+        .skip(columns::START_INSTRUCTION_FLAGS)
```
Comment (dlubarov) [inline evm/src/cpu/decode.rs:190]: A little more idiomatic IMO
```suggestion
    for &flag in &lv[columns::START_INSTRUCTION_FLAGS..columns::END_INSTRUCTION_FLAGS] {
```

**Review by dlubarov — APPROVED** [review]
LGTM 👍

File: evm/src/cpu/columns.rs (hunk 1 above)
Comment (nbgl) [reply evm/src/cpu/columns.rs:123]: Yeah, I usually prefer little-endian, but I thought big-endian was nicer here. (Also, the EVM is big-endian.)

File: evm/src/cpu/columns.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+pub const OPCODE_BITS: [usize; 8] = [
+    END_INSTRUCTION_FLAGS + 1,
```
[the remark this answers is not in the document]
Comment (nbgl) [reply evm/src/cpu/columns.rs:126]: Good catch!

File: evm/src/cpu/decode.rs (hunk 2 above)
Comment (nbgl) [reply evm/src/cpu/decode.rs:190]: Oh true lol. (The chain of iterators was a mindless copy/paste from a Clippy suggestion)

File: evm/src/cpu/columns.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+pub const IS_MSTORE8: usize = IS_MSTORE + 1;
+pub const IS_SLOAD: usize = IS_MSTORE8 + 1;
+pub const IS_SSTORE: usize = IS_SLOAD + 1;
+pub const IS_JUMP: usize = IS_SSTORE + 1;
+pub const IS_JUMPI: usize = IS_JUMP + 1;
+pub const IS_PC: usize = IS_JUMPI + 1;
+pub const IS_MSIZE: usize = IS_PC + 1;
+pub const IS_GAS: usize = IS_MSIZE + 1;
+pub const IS_JUMPDEST: usize = IS_GAS + 1;
+// Find the number of to push by reading the bottom 5 bits of the opcode.
```
Comment (unzvfu) [inline evm/src/cpu/columns.rs:75]: Nit: "of to" => "to"

Comment (nbgl) [reply evm/src/cpu/columns.rs]: Fixed! Thank you

**Review by dlubarov — APPROVED** [review]

**Pull request merged**